PR: https://github.com/odoo/odoo/pull/222007

From: a0d94b18b3ff8766fc0d50fd1223d6a6c7bdf522
From: std-odoo
Date: 2025-09-05 13:00:17

Breaking data model changes scores: del:1 + add:0, change matches:
-    google_gmail_authorization_code = fields.Char(string='Authorization Code', groups='base.group_system', copy=False)

Total Changes: 108

[IMP] google_gmail: allow using IAP instead of local credentials

Purpose
=======
Create a new IAP application that act like a proxy between the Gmail API
and the customer database. That way, the customer won't need to create
their API credentials and they will be able to use ours.

Technical flows
===============
Authentication process
*---------------------
1. User database make a request to `<IAP>/api/mail_oauth/1/gmail`
    - We check his subscription
    - We build the URL to redirect him to the Gmail login page
    - We return this URL
2. User browser is redirected to the URL we received from IAP
    - He fills his Gmail credential
3. User browser is redirected to `<IAP>/api/mail_oauth/1/gmail_callback`
   with the authorization_code
    - IAP used its API id / secret and the authorization_code
      received from the user browser, to get the refresh_token
      and the access_token
4. User browser is redirected to `<DB>/google_gmail/iap_confirm`
    - We check that the email used to login on Gmail is the same
      as the one set on the server
    - We save the refresh_token / access_token on the mail servers

Refresh token process
*--------------------
To refresh the tokens, the flow is way more simple

1. User database requests `<IAP>/api/mail_oauth/1/gmail_access_token` with
    - The database UUID
    - The refresh token
2. IAP check the subscription of the database UUID
3. IAP make a request to the Gmail API with
    - The refresh token he received from the user database
    - It's API secret / id
    - Then it returns the new tokens

Task-3061856

Part-of: odoo/odoo#222007
Related: odoo/upgrade#8181
Signed-off-by: Thibault Delavallee (tde) <tde@openerp.com>

================================= pseudo patch: =================================

--- a/addons/google_gmail/models/google_gmail_mixin.py
+++ b/addons/google_gmail/models/google_gmail_mixin.py
@@ -8,7 +8,7 @@ import requests
 
 from werkzeug.urls import url_encode
 
-from odoo import _, api, fields, models, tools
+from odoo import _, fields, models, tools
 from odoo.exceptions import AccessError, UserError
 from odoo.tools.urls import urljoin as url_join
 
@@ -27,31 +27,29 @@ class GoogleGmailMixin(models.AbstractModel):
     _description = 'Google Gmail Mixin'
 
     _SERVICE_SCOPE = 'https://mail.google.com/ https://www.googleapis.com/auth/userinfo.email'
+    _DEFAULT_GMAIL_IAP_ENDPOINT = 'https://gmail.api.odoo.com'
 
     active = fields.Boolean(default=True)
 
-    google_gmail_authorization_code = fields.Char(string='Authorization Code', groups='base.group_system', copy=False)
     google_gmail_refresh_token = fields.Char(string='Refresh Token', groups='base.group_system', copy=False)
     google_gmail_access_token = fields.Char(string='Access Token', groups='base.group_system', copy=False)
     google_gmail_access_token_expiration = fields.Integer(string='Access Token Expiration Timestamp', groups='base.group_system', copy=False)
     google_gmail_uri = fields.Char(compute='_compute_gmail_uri', string='URI', help='The URL to generate the authorization code from Google', groups='base.group_system')
 
-    @api.depends('google_gmail_authorization_code')
     def _compute_gmail_uri(self):
         Config = self.env['ir.config_parameter'].sudo()
         google_gmail_client_id = Config.get_param('google_gmail_client_id')
         google_gmail_client_secret = Config.get_param('google_gmail_client_secret')
+        is_configured = google_gmail_client_id and google_gmail_client_secret
         base_url = self.get_base_url()
 
-        redirect_uri = url_join(base_url, '/google_gmail/confirm')
-
-        if not google_gmail_client_id or not google_gmail_client_secret:
+        if not is_configured:
             self.google_gmail_uri = False
         else:
             for record in self:
                 google_gmail_uri = 'https://accounts.google.com/o/oauth2/v2/auth?%s' % url_encode({
                     'client_id': google_gmail_client_id,
-                    'redirect_uri': redirect_uri,
+                    'redirect_uri': url_join(base_url, '/google_gmail/confirm'),
                     'response_type': 'code',
                     'scope': self._SERVICE_SCOPE,
                     # access_type and prompt needed to get a refresh token
@@ -77,12 +75,57 @@ class GoogleGmailMixin(models.AbstractModel):
         if not self.env.is_admin():
             raise AccessError(_('Only the administrator can link a Gmail mail server.'))
 
-        if not self.google_gmail_uri:
+        email_normalized = tools.email_normalize(self[self._email_field])
+        if not email_normalized:
+            raise UserError(_('Please enter a valid email address.'))
+
+        Config = self.env['ir.config_parameter'].sudo()
+        google_gmail_client_id = Config.get_param('google_gmail_client_id')
+        google_gmail_client_secret = Config.get_param('google_gmail_client_secret')
+        is_configured = google_gmail_client_id and google_gmail_client_secret
+
+        if not is_configured:  # use IAP (see '/google_gmail/iap_confirm')
+            gmail_iap_endpoint = self.env['ir.config_parameter'].sudo().get_param(
+                'mail.server.gmail.iap.endpoint',
+                self._DEFAULT_GMAIL_IAP_ENDPOINT,
+            )
+            db_uuid = self.env['ir.config_parameter'].sudo().get_param('database.uuid')
+
+            # final callback URL that will receive the token from IAP
+            callback_params = url_encode({
+                'model': self._name,
+                'rec_id': self.id,
+                'csrf_token': self._get_gmail_csrf_token(),
+            })
+            callback_url = url_join(self.get_base_url(), f'/google_gmail/iap_confirm?{callback_params}')
+
+            iap_url = url_join(gmail_iap_endpoint, '/api/mail_oauth/1/gmail')
+            try:
+                response = requests.get(
+                    iap_url,
+                    params={'db_uuid': db_uuid, 'callback_url': callback_url},
+                    timeout=GMAIL_TOKEN_REQUEST_TIMEOUT)
+                response.raise_for_status()
+            except requests.exceptions.RequestException as e:
+                _logger.error('Can not contact IAP: %s.', e)
+                raise UserError(_('Oops, we could not authenticate you. Please try again later.'))
+
+            response = response.json()
+            if 'error' in response:
+                self._raise_iap_error(response['error'])
+
+            # URL on IAP that will redirect to Gmail login page
+            google_gmail_uri = response['url']
+
+        else:
+            google_gmail_uri = self.google_gmail_uri
+
+        if not google_gmail_uri:
             raise UserError(_('Please configure your Gmail credentials.'))
 
         return {
             'type': 'ir.actions.act_url',
-            'url': self.google_gmail_uri,
+            'url': google_gmail_uri,
             'target': 'self',
         }
 
@@ -106,8 +149,14 @@ class GoogleGmailMixin(models.AbstractModel):
         :return:
             access_token, access_token_expiration
         """
-        response = self._fetch_gmail_token('refresh_token', refresh_token=refresh_token)
+        Config = self.env['ir.config_parameter'].sudo()
+
+        google_gmail_client_id = Config.get_param('google_gmail_client_id')
+        google_gmail_client_secret = Config.get_param('google_gmail_client_secret')
+        if not google_gmail_client_id or not google_gmail_client_secret:
+            return self._fetch_gmail_access_token_iap(refresh_token)
 
+        response = self._fetch_gmail_token('refresh_token', refresh_token=refresh_token)
         return (
             response['access_token'],
             int(time.time()) + response['expires_in'],
@@ -144,6 +193,44 @@ class GoogleGmailMixin(models.AbstractModel):
 
         return response.json()
 
+    def _fetch_gmail_access_token_iap(self, refresh_token):
+        """Fetch the access token using IAP.
+
+        Make a HTTP request to IAP, that will make a HTTP request
+        to the Gmail API and give us the result.
+
+        :return:
+            access_token, access_token_expiration
+        """
+        gmail_iap_endpoint = self.env['ir.config_parameter'].sudo().get_param(
+            'mail.server.gmail.iap.endpoint',
+            self.env['google.gmail.mixin']._DEFAULT_GMAIL_IAP_ENDPOINT,
+        )
+        db_uuid = self.env['ir.config_parameter'].sudo().get_param('database.uuid')
+
+        response = requests.get(
+            url_join(gmail_iap_endpoint, '/api/mail_oauth/1/gmail_access_token'),
+            params={'refresh_token': refresh_token, 'db_uuid': db_uuid},
+            timeout=GMAIL_TOKEN_REQUEST_TIMEOUT,
+        )
+
+        if not response.ok:
+            _logger.error('Can not contact IAP: %s.', response.text)
+            raise UserError(_('Oops, we could not authenticate you. Please try again later.'))
+
+        response = response.json()
+        if 'error' in response:
+            self._raise_iap_error(response['error'])
+
+        return response
+
+    def _raise_iap_error(self, error):
+        errors = {
+            "not_configured": _("Gmail is not configured on IAP."),
+            "no_subscription": _("You don't have an active subscription."),
+        }
+        raise UserError(_('An error occurred: %s.', errors.get(error, error)))
+
     def _generate_oauth2_string(self, user, refresh_token):
         """Generate a OAuth2 string which can be used for authentication.
 

--- a/addons/google_gmail/models/ir_mail_server.py
+++ b/addons/google_gmail/models/ir_mail_server.py
@@ -39,7 +39,6 @@ class IrMail_Server(models.Model):
             self.smtp_encryption = 'starttls'
             self.smtp_port = 587
         else:
-            self.google_gmail_authorization_code = False
             self.google_gmail_refresh_token = False
             self.google_gmail_access_token = False
             self.google_gmail_access_token_expiration = False
